src/contracts/core/Slasher.sol

```diff
@@ -1,102 +1,368 @@
+                slashingRequest.id = requestId;
+            }
 
-    function getMiddlewareTimesIndexServeUntilBlock(address, uint32) external view returns (uint32) {}
+            // pending slashing rate is modified by the bips to modify * the slashable bips
+            // when divided by BIPS_FACTOR**2, this will give the actual proportion of the
+            // stake is being modified. this is done for accuracy when slashable bips are small
+            slashingRequest.slashingRate = uint64(
+                int64(slashingRequest.slashingRate)
+                    + int64(bipsToModify)
+                        * int64(uint64(operatorSetManager.getSlashableBips(operator, operatorSet, strategy, epoch)))
+            );
+            slashingRequests[operator][strategy][epoch] = slashingRequest;
+        }
 
-    function operatorWhitelistedContractsLinkedListSize(address) external view returns (uint256) {}
+        emit RequestedBipsToSlashModified(epoch, operator, operatorSet, strategies, bipsToModify);
+    }
 
-    function operatorWhitelistedContractsLinkedListEntry(
-        address,
-        address
-    ) external view returns (bool, uint256, uint256) {}
+    /// @notice Returns the epoch for lookup in the slashing request history and whether
+    /// there were any slashing requests that happened before or at the given epoch
+    function _getLookupEpoch(address operator, IStrategy strategy, uint32 epoch) internal view returns (uint32, bool) {
+        // TODO DONE?: note the edge case of 0th epoch; need to make sure it's clear how it should be handled
+        uint32 epochForLookup;
+        bool found;
+        for (uint256 i = slashedEpochHistory[operator][strategy].length; i > 0; --i) {
+            epochForLookup = slashedEpochHistory[operator][strategy][i - 1];
+            if (epochForLookup <= epoch) {
+                found = true;
+                break;
+            }
+        }
+        return (epochForLookup, found);
+    }
 
-    function whitelistedContractDetails(
-        address,
-        address
-    ) external view returns (MiddlewareDetails memory) {}
+    /// @notice the withdrawability condition for a certain epoch that a withdrawal was slashable until is
+    /// that all slashings executable at that epoch have been executed
+    /// @dev slashingRequests[operator][strategy][epochForLookup] must be set, use _getLookupEpoch
+    function _canWithdraw(address operator, IStrategy strategy, uint32 epoch) internal view returns (bool) {
+        return slashingRequests[operator][strategy][epoch].id
+            <= slashingRequestIds[operator][strategy].lastExecutedSlashingRequestId;
+    }
 
+    function _hashOperatorSet(IOperatorSetManager.OperatorSet memory operatorSet) internal pure returns (bytes32) {
+        IOperatorSetManager.OperatorSet[] memory operatorSetArray = new IOperatorSetManager.OperatorSet[](1);
+        operatorSetArray[0] = operatorSet;
+        return keccak256(abi.encode(operatorSetArray));
```

Comment:
is there a reason to hash a length-1 array of the struct instead of just hashing the struct itself?